File: src/main/java/com/easy/kotlin/ReflectionDemo.java
/*
 * Copyright (c) 2017. Lorem ipsum dolor sit amet, consectetur adipiscing elit.
 * Morbi non lorem porttitor neque feugiat blandit. Ut vitae ipsum eget quam lacinia accumsan.
 * Etiam sed turpis ac ipsum condimentum fringilla. Maecenas magna.
 * Proin dapibus sapien vel ante. Aliquam erat volutpat. Pellentesque sagittis ligula eget metus.
 * Vestibulum commodo. Ut rhoncus gravida arcu.
 */

package com.easy.kotlin;

import java.lang.annotation.Annotation;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;

interface StudentService<T> {
    List<T> findStudents(String name, Integer age);
}

public class ReflectionDemo {

    public static void main(String[] args) {
        StudentServiceImpl studentService = new StudentServiceImpl();
        studentService.save(new Student("Bob", 20));
        studentService.findStudents("Jack", 20);

        // 反射API调用示例
        final Class<? extends StudentServiceImpl> studentServiceClass = studentService.getClass();
        Class<?>[] classes = studentServiceClass.getDeclaredClasses();
        Annotation[] annotations = studentServiceClass.getAnnotations();
        ClassLoader classLoader = studentServiceClass.getClassLoader(); // Returns the class loader for the class
        Field[] fields = studentServiceClass.getDeclaredFields(); // 获取类成员变量
        Method[] methods = studentServiceClass.getDeclaredMethods(); // 获取类成员方法
        try {
            methods[0].getName(); // save
            methods[0].invoke(studentService, "Jack",20);
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        }
    }

}

class StudentServiceImpl extends BaseService<Student> implements StudentService<Student> {

    public List<Student> findStudents(String name, Integer age) {
        return Arrays.asList(new Student[] {new Student("Jack", 20), new Student("Rose", 20)});
    }

    @Override
    public int save(Student student) {
        return 0;
    }
}

abstract class BaseService<T> {
    abstract int save(T t);
}

class Student {

    String name;
    Integer age;

    public Student(String name, Integer age) {
        this.name = name;
        this.age = age;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Integer getAge() {
        return age;
    }

    public void setAge(Integer age) {
        this.age = age;
    }
}

File: src/main/kotlin/com/easy/kotlin/reflection/ReflectionDemo.kt
/*
 * Copyright (c) 2017. Lorem ipsum dolor sit amet, consectetur adipiscing elit.
 * Morbi non lorem porttitor neque feugiat blandit. Ut vitae ipsum eget quam lacinia accumsan.
 * Etiam sed turpis ac ipsum condimentum fringilla. Maecenas magna.
 * Proin dapibus sapien vel ante. Aliquam erat volutpat. Pellentesque sagittis ligula eget metus.
 * Vestibulum commodo. Ut rhoncus gravida arcu.
 */

package com.easy.kotlin.reflection

import java.lang.reflect.ParameterizedType
import kotlin.reflect.KTypeParameter
import kotlin.reflect.full.declaredFunctions

open class BaseContainer<T>

class Container<T : Comparable<T>> : BaseContainer<Int> {
    var elements: MutableList<T>

    constructor(elements: MutableList<T>) {
        this.elements = elements
    }

    fun sort(): Container<T> {
        elements.sort()
        return this
    }

    override fun toString(): String {
        return "Container(elements=$elements)"
    }
}


fun fooContainer() {
    val container = Container(mutableListOf<Int>(1, 3, 2, 5, 4, 7, 6))
    println(container)
    container.sort()
    println(container)

    val kClass = container::class // 获取KClass对象

    val jkCLass = kClass.java
    println(jkCLass)
//    val staticKClass = Container::class
//    println(staticKClass)

    val typeParameters = kClass.typeParameters // 获取类型参数typeParameters信息，也即泛型信息

    val kTypeParameter: KTypeParameter = typeParameters[0]
    println(kTypeParameter.isReified) // false
    println(kTypeParameter.name) // T
    println(kTypeParameter.upperBounds) // [kotlin.Comparable<T>]
    println(kTypeParameter.variance) // INVARIANT

    val constructors = kClass.constructors
    for (KFunction in constructors) {
        KFunction.parameters.forEach {
            val name = it.name
            val type = it.type
            println("name = ${name}") // elements
            println("type = ${type}") // kotlin.collections.MutableList<T>
            for (KTypeProjection in type.arguments) {
                println(KTypeProjection.type) // T
            }
        }
    }

    kClass.declaredFunctions
    kClass.members
    kClass.typeParameters

    val jClass = container.javaClass // 获取Java Class对象

    val parameterizedType = jClass.genericSuperclass as ParameterizedType // ParameterizedType represents a parameterized type such a Collection<String>
    val actualTypeArguments = parameterizedType.actualTypeArguments
    for (type in actualTypeArguments) {
        val typeName = type.typeName
        println("typeName = ${typeName}")
    }

    val rawType = parameterizedType.rawType
    println("rawType = ${rawType}")
}

fun <A, B, C> compose(f: (B) -> C, g: (A) -> B): (A) -> C {
    return { x -> f(g(x)) }
}

fun isOdd(x: Int) = x % 2 != 0
fun length(s: String) = s.length


class A<T>
open class C<T>
class B<T> : C<Int>()

fun fooA() {
    // 无法在此处获得运行时 T 的具体类型！！！
    val parameterizedType = A<Int>()::class.java.genericSuperclass as ParameterizedType
    val actualTypeArguments = parameterizedType.actualTypeArguments
    for (type in actualTypeArguments) {
        val typeName = type.typeName
        println("typeName = ${typeName}")
    }
}

fun fooB() {
    // 当继承了父类 C<Int> 的时候，在此处获得运行时 genericSuperclass T 的具体类型
    val parameterizedType = B<Int>()::class.java.genericSuperclass as ParameterizedType
    val actualTypeArguments = parameterizedType.actualTypeArguments
    for (type in actualTypeArguments) {
        val typeName = type.typeName
        println("typeName = ${typeName}")
    }
}


fun main(args: Array<String>) {
    fooContainer()
//    fooA()
//    fooB()

//    val nums = listOf(1, 2, 3)
//    val filteredNums = nums.filter(::isOdd)
//    println(filteredNums)
//
//    testReflectProperty()
//
//    testBindRef()

}


var one = 1
fun testReflectProperty() {
    println(::one.get()) // 1
    ::one.set(2)
    println(one)         // 2
}


fun testBindRef() {
    val digitRegex = "\\d+".toRegex()
    digitRegex.matches("7") // true
    digitRegex.matches("6") // true
    digitRegex.matches("5") // true
    digitRegex.matches("X") // false

    val isDigit = digitRegex::matches // 绑定函数 isNumber = numberRegex::matches
    isDigit("7")// true
    isDigit("6")// true
    isDigit("5")// true
    isDigit("X")// true
}



